### PR TITLE
refactor(tree2): Delta format based on detach/attach model

### PR DESCRIPTION
**Pull request by yann-achard-MS**
Refactors the Delta format to adopt a model based on nodes being detached and attached.

This simplifies the format and makes it more homogenous (moves and insertions/removals) now leverage the same abstractions.
It also allows the format to represent a wider range of changes. For example, it allows the representation of a node being detached for a move, but never attached.
The format no longer represents edits to removed content inline within the field that last contained that content.

### CODE REVIEW

File: experimental/dds/tree2/src/core/tree/delta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/delta.ts:15]: I don't understand the example. There's also an extra "represents" at the end of the line.

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/tree/delta.ts:15]: Would it be clearer if I called out tiebreaking information not being included (since that's what this example is getting at)?

File: experimental/dds/tree2/src/core/tree/delta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/delta.ts:150]: This name is awkward. Could we just call this `FieldMap`? Can the existing `FieldMap` alias just be inlined?

File: experimental/dds/tree2/src/core/tree/delta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/delta.ts:185]: Why is this field not marked readonly?

File: experimental/dds/tree2/src/core/tree/deltaUtil.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/deltaUtil.ts:13]: Not related to this PR, but we should probably do something like freezing this or changing the types to make it less likely that this is ever accidentally mutated. If this is only used in tests, we should probably move it to a test file to make it less likely to cause bugs in production.

File: experimental/dds/tree2/src/core/tree/deltaUtil.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/deltaUtil.ts:49]:
```suggestion
	const mark: Mark = { count: 1, attach: buildId };
	if (detachId !== undefined) {
		mark.detach = detachId;
	}
```

Seems more natural to me.

File: experimental/dds/tree2/src/core/tree/delta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/delta.ts:188]: I think we should make the grammatical case consistent for these field names. How about build -> built and rename -> renamed?

Comment (alex-pardes) [reply experimental/dds/tree2/src/core/tree/delta.ts:188]: I guess I see why they are in different cases, but maybe we should make the names more verbose. Maybe something like `attachedChanges`, `detachedChanges`, `builds`, and `renames`?

File: experimental/dds/tree2/src/core/tree/delta.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/tree/delta.ts:185]: `MarkList` was defined elsewhere as readonly. I've now inlined the definition for consistency.

File: experimental/dds/tree2/src/core/tree/deltaUtil.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/tree/deltaUtil.ts:13]: The types do prevent changes already for both of these.

File: experimental/dds/tree2/src/core/tree/delta.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/tree/delta.ts:195]: I was worried this would be confusing. I find `attachedChanges` a bit confusing as well since it's not the changes that are attached/detached (`attachedNodesChanges` would be clearer but it's getting rather long which will make the code formatting ugly). How about `inlineChanges` and `rootChanges`?

File: experimental/dds/tree2/src/core/tree/visitDelta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/visitDelta.ts:305]:
```suggestion
 * Performs the following:
```

File: experimental/dds/tree2/src/core/tree/visitDelta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/core/tree/visitDelta.ts:400]:
```suggestion
							config.detachedFieldIndex.getEntry(offsetAttachId);
```

The entry should always be defined here, right?

File: experimental/dds/tree2/src/feature-libraries/default-field-kinds/optionalField.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/feature-libraries/default-field-kinds/optionalField.ts:498]: Isn't this incorrectly ignoring changes to detached nodes?

File: experimental/dds/tree2/src/feature-libraries/deltaUtils.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/feature-libraries/deltaUtils.ts:72]: The ungrammatical "represented with by" is used twice here.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:27]: It looks like this line could be moved to the end of the function.

File: experimental/dds/tree2/src/core/tree/delta.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/tree/delta.ts:103]: Yeah, I tried splitting this type into 3 or 4 to call out the various cases, but it make the whole thing very cumbersome.
Do you think adding some comments would be good enough?

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:53]: Is this supposed to say "unexpected"?

File: experimental/dds/tree2/src/core/tree/visitDelta.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/core/tree/visitDelta.ts:256]: No:
- detachPassRoots is populated in the detach pass then consumed in the detach pass
- attachPassRoots is populated in both passes and consumed in the attach pass

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:88]: Is this waiting until we change our merge behavior to apply changes to detached trees?

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts
Comment (alex-pardes) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:84]: I think this case should never be hit as we've already checked that `mark` changes the cell content.

File: experimental/dds/tree2/src/feature-libraries/default-field-kinds/optionalField.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/default-field-kinds/optionalField.ts:498]: I would say it is correctly doing so. We're not allowing those yet (that will be done as a separate PR).

File: experimental/dds/tree2/src/feature-libraries/deltaUtils.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/deltaUtils.ts:72]: Oh dear, what happended here?!

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:53]: Yes.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/sequenceFieldToDelta.ts:88]: Yes

**Review by alex-pardes — APPROVED** [review]

**Pull request merged**
